[InstCombine] Remove Store when its Ptr is removable

ParkHanbum opened the pull request:
If the `store` target is OneUse and the user is `AddrSpaceCastInst`, proper processing was not performed. Therefore, when processing `Alloca` in the second process, this storage space was deleted and a problem occurred.

This patch removes the `store` when the store target is OneUse and the user is `AddrSpaceCastInst`.

This is a sample IR a crash occurred. crafted by @arsenm 
```
define void @test(ptr align 8 %arg) {
bb:
  %i = alloca ptr, align 8, addrspace(5)
  %i1 = addrspacecast ptr addrspace(5) %i to ptr
  store ptr %arg, ptr %i1, align 8
  %i2 = load ptr, ptr %i1, align 8
  store ptr %i2, ptr null, align 8
  ret void
}
```

RUN RECORD:

$ bllvm0/bin/clang --version
clang version 18.0.0git (git@github.com:ParkHanbum/llvm-project.git 0633226612a37933ebd41e523bb9383d80b4cf6a)
$ bllvm0/bin/opt --passes=instcombine crash.ll -debug
```
INSTCOMBINE ITERATION #1 on test
ADD:   ret void
ADD:   store ptr %i2, ptr null, align 8
ADD:   %i2 = load ptr, ptr %i1, align 8
ADD:   store ptr %arg, ptr %i1, align 8
ADD:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
ADD:   %i = alloca ptr, align 8, addrspace(5)
IC: Visiting:   %i = alloca ptr, align 8, addrspace(5)
IC: Visiting:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
IC: Visiting:   store ptr %arg, ptr %i1, align 8
IC: Visiting:   %i2 = load ptr, ptr %i1, align 8
IC: Replacing   %i2 = load ptr, ptr %i1, align 8
    with ptr %arg
IC: Mod =   %i2 = load ptr, ptr %i1, align 8
    New =   %i2 = load ptr, ptr %i1, align 8
IC: ERASE   %i2 = load ptr, ptr %i1, align 8
ADD DEFERRED:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
ADD DEFERRED:   store ptr %arg, ptr %i1, align 8
ADD:   store ptr %arg, ptr %i1, align 8
ADD:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
IC: Visiting:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
IC: Visiting:   store ptr %arg, ptr %i1, align 8
IC: Visiting:   store ptr %arg, ptr null, align 8
IC: Mod =   store ptr %arg, ptr null, align 8
    New =   store ptr poison, ptr null, align 8
ADD:   store ptr poison, ptr null, align 8
IC: Visiting:   store ptr poison, ptr null, align 8
IC: Visiting:   ret void


INSTCOMBINE ITERATION #2 on test
ADD:   ret void
ADD:   store ptr poison, ptr null, align 8
ADD:   store ptr %arg, ptr %i1, align 8
ADD:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
ADD:   %i = alloca ptr, align 8, addrspace(5)
IC: Visiting:   %i = alloca ptr, align 8, addrspace(5)
IC: Replacing   %i1 = addrspacecast ptr addrspace(5) %i to ptr
    with ptr poison
IC: ERASE   %i1 = addrspacecast ptr addrspace(5) %i to ptr
ADD DEFERRED:   %i = alloca ptr, align 8, addrspace(5)
IC: ERASE   store ptr %arg, ptr poison, align 8
IC: ERASE   %i = alloca ptr, align 8, addrspace(5)
IC: Visiting:   store ptr poison, ptr null, align 8
IC: Visiting:   ret void
LLVM ERROR: Instruction Combining did not reach a fixpoint after 1 iterations
```

ANALYZE : 

This problem seems to come from incomplete processing of the `addrspacecast` instruction.
In the first instcombine process, after the load instruction is removed as follows:
```
IC: Replacing   %i2 = load ptr, ptr %i1, align 8
    with ptr %arg
IC: Mod =   %i2 = load ptr, ptr %i1, align 8
    New =   %i2 = load ptr, ptr %i1, align 8
IC: ERASE   %i2 = load ptr, ptr %i1, align 8
ADD DEFERRED:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
ADD DEFERRED:   store ptr %arg, ptr %i1, align 8
ADD:   store ptr %arg, ptr %i1, align 8
ADD:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
IC: Visiting:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
IC: Visiting:   store ptr %arg, ptr %i1, align 8
IC: Visiting:   store ptr %arg, ptr null, align 8
IC: Mod =   store ptr %arg, ptr null, align 8
    New =   store ptr poison, ptr null, align 8
ADD:   store ptr poison, ptr null, align 8
IC: Visiting:   store ptr poison, ptr null, align 8
IC: Visiting:   ret void
```
In the processing the first instruction `alloc` in the second instcombine process, `isAllocSiteRemovable` inside the `visitAllocSite` function is called.
```
ADD:   ret void
ADD:   store ptr poison, ptr null, align 8
ADD:   store ptr %arg, ptr %i1, align 8
ADD:   %i1 = addrspacecast ptr addrspace(5) %i to ptr
ADD:   %i = alloca ptr, align 8, addrspace(5)
IC: Visiting:   %i = alloca ptr, align 8, addrspace(5)
IC: Replacing   %i1 = addrspacecast ptr addrspace(5) %i to ptr
    with ptr poison
IC: ERASE   %i1 = addrspacecast ptr addrspace(5) %i to ptr
ADD DEFERRED:   %i = alloca ptr, align 8, addrspace(5)
```
In this function, the `addrspacecast` and `store` instructions are removed.

This is the processing that should have occurred when the `load` instruction was removed. However, addrspacecast does not do this.

arsenm commented on llvm/test/Transforms/InstCombine/store.ll:
```diff
@@ -345,6 +345,22 @@ define void @store_to_readonly_noalias(ptr readonly noalias %0) {
+  %i1 = addrspacecast ptr addrspace(5) %i to ptr
+  store ptr %arg, ptr %i1, align 8
+  %i2 = load ptr, ptr %i1, align 8
+  store ptr %i2, ptr null, align 8
```
Use something more meaningful than an undefined store to null?

ParkHanbum replied:
ok, I'll try to figure it out.

arsenm commented on llvm/test/Transforms/InstCombine/store.ll:
```diff
@@ -345,6 +345,22 @@ define void @store_to_readonly_noalias(ptr readonly noalias %0) {
 }
 
+; if ptr which store to is addrspacecast, and its target is alloca. remove it.
+define void @src_store_oneuse_addrspaceast_alloca(ptr align 8 %arg) {
```
Should run with the max iterations flag to make sure that problem is fixed?

ParkHanbum replied:
Is this patch not enough to fix this issue? I thought it would be enough to solve the problem of `reduce-instcombine-fixpoint-error.ll.txt`.

arsenm replied:
I mean I don't remember the exact iteration status. Without this patch, and without additional pass arguments, do you see the verifier error in this test as-is?

ParkHanbum replied:
I see. I was worried that I might have misunderstood and written a patch.

arsenm replied:
I just checked. I see the fix point error without any flags, so it's fine.

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineLoadStoreAlloca.cpp:
```diff
@@ -1385,9 +1385,10 @@ Instruction *InstCombinerImpl::visitStoreInst(StoreInst &SI) {
-      if (isa<AllocaInst>(GEP->getOperand(0))) {
-        if (GEP->getOperand(0)->hasOneUse())
+    if (isa<GetElementPtrInst>(Ptr) || isa<AddrSpaceCastInst>(Ptr)) {
+      Instruction *PtrI = dyn_cast<Instruction>(Ptr);
```
Unchecked dyn_cast, use cast<>

arsenm commented on llvm/test/Transforms/InstCombine/store.ll:
```diff
@@ -345,6 +345,22 @@ define void @store_to_readonly_noalias(ptr readonly noalias %0) {
+; CHECK-NEXT:    ret void
+;
+bb:
+  %i = alloca ptr, align 8, addrspace(5)
+  %i1 = addrspacecast ptr addrspace(5) %i to ptr
+  store ptr %arg, ptr %i1, align 8
```
This test case doesn't contain a getelementptr. I don't see how anything could have gone wrong previously? Either way, should have a test with and without GEP

ParkHanbum replied:
to summarize, do you mean that a test with GEP is also necessary? but imo, this issue have not related with GEP. I would appreciate it if you could tell me why you are referring to GEP.

arsenm replied:
I mean you changed the handling under the dyn_cast<GetElementPtrInst>. As such I don't see how this could possibly fix anything for the AddrSpaceCastInst case. There's nothing special about AddrSpaceCast specifically. I assume the same would manifest for any transitive pointer user, so I don't understand how this is avoiding the issue

ParkHanbum replied:
@arsenm Now I understand!! I think it would be solve to implement a function like `visitallocsite` in the `alloca` instruction. thanks!!

ParkHanbum commented:
I left a comment about the issue explaining how to fix it. I will close this pull request as it seems unnecessary until this issue is resolved. may I? @arsenm

ParkHanbum committed:
[InstCombine] Remove Store when its Ptr is removable (#68120)

Previously, if the `Store`'s `Ptr` was `Alloca` and satisfy OneUse()
it would be removed. However, if `Alloca` was overlapped and
referenced by instructions such as `GEP` or `AddrSpaceCast`, it could
not be removed and an error occurred as a result.

This patch fixes this by having the `Store`'s `Ptr` find out
if the target is `Alloca`, and remove the `Store` if possible.

nikic reviewed (changes requested):
This issue is Won't Fix as far as I'm concerned. Reaching a fix-point in *all* cases is an explicit non-goal -- the verification exists to catch bugs like failing to add things back to the worklist. You should not perform contortions to reach the fixpoint, unless there is phase-ordering proof that doing so would be useful.

ParkHanbum commented:
When the Store's Ptr is Alloca, we changed it to determine whether it can be removed by calling isAllocSiteRemovable, and added some tests. This solve the current problem.

However, there is still room for improvement, and this has been left as a comment on the issue.

arsenm commented:
I think this is resolvable by just using stripPointerCasts in visitStoreInst when it's looking for the alloca/GEP source

arsenm commented:
Put up a WIP https://github.com/arsenm/llvm-project/tree/issue68120-worklist-issue

nikic commented:
> > This issue is Won't Fix as far as I'm concerned. Reaching a fix-point in _all_ cases is an explicit non-goal -- the verification exists to catch bugs like failing to add things back to the worklist.
> 
> If this is the case, I think the current behavior of error-by-default isn't really reasonable. You should be able to run instcombine standalone without triggering errors like this

Our requirement is that fix-point verification is enabled by default when writing instcombine tests, to prevent accidental introduction of worklist management bugs. The way we currently do this is to make `-passes=instcombine` enable fix-point verification, as that's what we use in tests. When instantiating `InstCombinePass()` from C++ code we disable it by default, as the verification should not be enabled in any production use.

Those are basically the constraints here -- happy to hear alternative ways we could have this "verify in tests, don't verify in production" behavior. Possibly the error message just needs to be improved to make it clear that this is not necessary a problem?

> Put up a WIP https://github.com/arsenm/llvm-project/tree/issue68120-worklist-issue

This will miscompile cases if there are multiple uses along the stripped chain, I think?

arsenm commented:
> Possibly the error message just needs to be improved to make it clear that this is not necessary a problem?

I think the message definitely should be improved. Downgrade the message to warning?

> > Put up a WIP https://github.com/arsenm/llvm-project/tree/issue68120-worklist-issue
> 
> This will miscompile cases if there are multiple uses along the stripped chain, I think?

I started guarding against these, but so far I haven't had much luck avoiding the second iteration in the mixed multiple use cases.
